Add DNS support to proxy protocol query interceptor

mpenick opened the pull request:
This adds functionality to `ProxyProtocolQueryInterceptor` to populate `system.peers` with the A records of a proxy that lives in front of Stargate. This completes the functionality that allows Stargate to live behind a proxy e.g. Envoy, HA proxy, etc. and use the proxy's public IP addresses in `system.local` and `system.peers`. It also monitors the current state of the A records and if they change it generates the appropriate topology events for the CQL protocol.

I've added a good amount of unit tests around `ProxyProtocolQueryInterceptor`. In the future, it will make sense to create a way to simulate a proxy that implements the proxy protocol in an integration test. That might be out of the scope of this ticket for now.

This also fixes https://github.com/stargate/stargate/issues/250. The persistence layer now deals with `InetAddress` and the cql layer adds the appropriate port, using `InetAddressAndPort`, to the events on the way out.

dougwettlaufer commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 10
           

Complexity increasing per file
==============================
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptorTest.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java  21
         

Clones added
============
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/Conversion.java  1
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/Conversion.java  1
- persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/Conversion.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6334016&bid=20577739)

dougwettlaufer commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 8
           

Complexity increasing per file
==============================
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java  21
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptorTest.java  3
         

Clones added
============
- persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/Conversion.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/Conversion.java  1
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/Conversion.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6334016&bid=20577739)

mpenick commented on cql/src/main/java/org/apache/cassandra/stargate/transport/internal/Server.java:
```diff
@@ -626,37 +628,43 @@ private void send(Event event) {
-    public void onDown(InetAddressAndPort endpoint) {
-      onStatusChange(
-          endpoint, Event.StatusChange.nodeDown(server.persistence.getNativeAddress(endpoint)));
+    private InetAddressAndPort withPort(InetAddress endpoint) {
```
This adds the native protocol port to the event's address.

olim7t commented on cql/src/main/java/org/apache/cassandra/stargate/transport/internal/Server.java:
```diff
@@ -626,37 +628,43 @@ private void send(Event event) {
+          endpoint,
+          server.useSSL
+              ? TransportDescriptor.getNativeTransportPortSSL()
+              : TransportDescriptor.getNativeTransportPort());
```
Those are the ports of the current instance though. Are we sure that the remote node will use the same ones?

olim7t commented on persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java:
```diff
@@ -27,15 +44,68 @@
+    if (Strings.isNullOrEmpty(ttl)) {
+      logger.info(
+          "DNS cache TTL (property \"networkaddress.cache.ttl\") not explicitly set. Setting to 60 seconds.");
+      Security.setProperty("networkaddress.cache.ttl", "60");
```
Ah nice, was going to mention that 👍

olim7t commented on persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java:
```diff
@@ -50,91 +120,158 @@ public void initialize() {}
+  private static class DefaultResolver implements Resolver {
+    @Override
+    public Set<InetAddress> resolve(String name) throws UnknownHostException {
+      return Arrays.stream(InetAddress.getAllByName(name)).collect(Collectors.toSet());
```
You mentioned in the PR description:
> It also monitors the current state of the A records and if they change it generates the appropriate topology events for the CQL protocol

Where does that happen?

mpenick replied:
Here: https://github.com/stargate/stargate/pull/267/files#diff-91bef8000571404fe1179e3c6e3e6cd6aed7fa2f9ba59ad198cd44e136dce632R168-R180

olim7t replied:
Ok, I don't know how I could miss it, thanks.

dougwettlaufer commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 11
           

Complexity increasing per file
==============================
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptorTest.java  3
- stargate-starter/src/main/java/io/stargate/starter/Starter.java  1
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/BaseDseTest.java  2
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java  21
         

Clones added
============
- persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/Conversion.java  1
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/Conversion.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/Conversion.java  1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6334016&bid=20577739)

mpenick commented on cql/src/main/java/org/apache/cassandra/stargate/transport/internal/Server.java:
```diff
@@ -626,37 +628,43 @@ private void send(Event event) {
+    public void onJoinCluster(InetAddress endpoint) {
+      InetAddressAndPort endpointWithPort = withPort(endpoint);
+      if (server.isRpcReady.get()) endpointsPendingJoinedNotification.add(endpointWithPort);
+      else onTopologyChange(endpointWithPort, Event.TopologyChange.newNode(endpointWithPort));
```
Thanks, you're right. There's some fundamental broken stuff. Esp. with the event too.

dougwettlaufer commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 5
- Added 11
           

Complexity increasing per file
==============================
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/BaseDseTest.java  2
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptorTest.java  3
- stargate-starter/src/main/java/io/stargate/starter/Starter.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java  21
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6334016&bid=20577739)

dougwettlaufer commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 12
- Added 15
           

Complexity increasing per file
==============================
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java  21
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  2
- persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  3
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/BaseDseTest.java  2
- stargate-starter/src/main/java/io/stargate/starter/Starter.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  2
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptorTest.java  3
         

Clones added
============
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  1
         

Clones removed
==============
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/EventListenerWrapper.java  -1
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/EventListenerWrapper.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6334016&bid=20577739)

dougwettlaufer commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 12
- Added 15
           

Complexity increasing per file
==============================
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  2
- persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  3
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  2
- stargate-starter/src/main/java/io/stargate/starter/Starter.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java  21
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/BaseDseTest.java  2
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptorTest.java  3
         

Clones added
============
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  1
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  1
         

Clones removed
==============
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/EventListenerWrapper.java  -1
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/EventListenerWrapper.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6334016&bid=20577739)

dougwettlaufer commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 12
- Added 14
           

Complexity increasing per file
==============================
- stargate-starter/src/main/java/io/stargate/starter/Starter.java  1
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/BaseDseTest.java  2
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java  21
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  2
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  2
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptorTest.java  3
- persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  3
         

Clones added
============
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  1
         

Clones removed
==============
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/EventListenerWrapper.java  -1
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/EventListenerWrapper.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6334016&bid=20577739)

dougwettlaufer commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 12
- Added 5
           

Complexity increasing per file
==============================
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptorTest.java  3
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/BaseDseTest.java  2
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  2
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  2
- persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  3
- stargate-starter/src/main/java/io/stargate/starter/Starter.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java  21
         

Clones added
============
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  1
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  1
         

Clones removed
==============
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/EventListenerWrapper.java  -1
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/EventListenerWrapper.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6334016&bid=20577739)

dougwettlaufer commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 35
- Added 12
           

Complexity increasing per file
==============================
- persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  9
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  11
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  6
- stargate-starter/src/main/java/io/stargate/starter/Starter.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java  21
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptorTest.java  3
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/BaseDseTest.java  2
         

Complexity decreasing per file
==============================
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/StargateSystemKeyspace.java  -13
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java  -7
+ persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java  -9
         

Clones added
============
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  2
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  1
         

Clones removed
==============
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/StargateSystemKeyspace.java  -1
+ persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  -1
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java  -1
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/EventListenerWrapper.java  -1
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/EventListenerWrapper.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6334016&bid=20577739)

mpenick commented on cql/src/main/java/org/apache/cassandra/stargate/transport/internal/Server.java:
```diff
@@ -622,41 +615,39 @@ private void send(Event event) {
-
     @Override
-    public void onJoinCluster(InetAddressAndPort endpoint) {
-      if (server.persistence.isRpcReady(endpoint)) endpointsPendingJoinedNotification.add(endpoint);
```
Move this RPC ready and native address logic into the default interceptors.

dougwettlaufer commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 35
- Added 12
           

Complexity increasing per file
==============================
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java  21
- stargate-starter/src/main/java/io/stargate/starter/Starter.java  1
- persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  9
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptorTest.java  3
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  6
- persistence-dse-6.8/src/test/java/io/stargate/db/dse/impl/BaseDseTest.java  2
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  11
         

Complexity decreasing per file
==============================
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/StargateSystemKeyspace.java  -13
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java  -7
+ persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java  -9
         

Clones added
============
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  2
         

Clones removed
==============
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/EventListenerWrapper.java  -1
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/StargateSystemKeyspace.java  -1
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/EventListenerWrapper.java  -1
+ persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  -1
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6334016&bid=20577739)

mpenick commented:
Tested manually using: https://gist.github.com/mpenick/d1baf06957d0df0870935e70a8470571 and https://github.com/datastax/cpp-driver/blob/2.15.3/examples/host_listener/host_listener.c 

Added and removed multiple stargate node verifying that the correct events were created. I plan to follow up with automated tests in https://github.com/stargate/stargate/issues/25 after we add ability to add/remove stargate nodes in tests.

tomekl007 commented on cql/src/main/java/io/stargate/cql/CqlActivator.java:
```diff
@@ -66,19 +66,17 @@ private synchronized void maybeStartService(
+      return;
     }
-
+    log.info(("Stopping CQL"));
```
nit: double `((`

tomekl007 commented on cql/src/main/java/io/stargate/cql/CqlActivator.java:
```diff
@@ -66,19 +66,17 @@ private synchronized void maybeStartService(
-    if (c != null) {
-      log.info(("Stopping CQL"));
-      c.stop();
+    if (cql != null) { // Not started
```
how do we know that it is not started in that case? I would suspect that if `cql == null` then not started.

tomekl007 commented on cql/src/main/java/org/apache/cassandra/stargate/transport/internal/Server.java:
```diff
@@ -622,41 +615,39 @@ private void send(Event event) {
-          endpoint, Event.StatusChange.nodeDown(server.persistence.getNativeAddress(endpoint)));
+    private InetAddressAndPort addPort(InetAddress endpoint, int port) {
+      return InetAddressAndPort.getByAddressOverrideDefaults(
+          endpoint, port == NO_PORT ? server.socket.getPort() : port);
```
so it basically takes the default port if there is no explicit port, right?
nit: maybe we could extract `port == NO_PORT ? server.socket.getPort() : port` to a separate method: `getPortOrDefault`.

mpenick replied:
Good suggestion.

tomekl007 commented on persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java:
```diff
@@ -94,37 +114,179 @@ public void beforeChange(
+  public void onRestart(InetAddress endpoint, EndpointState state) {}
+
+  private boolean joinCluster(InetAddress endpoint, EndpointState state) {
+    if (!liveStargateNodes.add(endpoint)) {
```
when this situation can occur?

mpenick replied:
This can happen in `onChange()` when the nodes already been added. In that case it should update the single endpoint's application state.

tomekl007 commented on persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java:
```diff
@@ -98,37 +117,186 @@ public void beforeChange(
+    StargateSystemKeyspace.updatePeerInfo(
+        endpoint,
+        "tokens",
+        Collections.singleton(DatabaseDescriptor.getPartitioner().getMinimumToken().toString()));
```
is it safe to use `toString()`? Maybe we should have a dedicated method for parsing to string?

mpenick replied:
You're right, I'm now using `TokenFactory.toString(Token)` which is the correct way of doing this.

tomekl007 commented on persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java:
```diff
@@ -94,37 +114,179 @@ public void beforeChange(
+  private void notifyJoinCluster(InetAddress endpoint) {
+    InetAddress nativeAddress = getNativeAddress(endpoint);
+    for (EventListener listener : listeners) {
+      listener.onJoinCluster(nativeAddress, EventListener.NO_PORT);
```
maybe we could get rid of the extra `NO_PORT` case, by passing the `InetSocketAddress` everywhere?
We could construct the `InetSocketAddress` based on the `InetAddress` and `server.socket.getPort()` when the `InetAddress` is received (at the higher level), wdyt?

tomekl007 commented on persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java:
```diff
@@ -94,37 +114,179 @@ public void beforeChange(
+        }
+        break;
+      case SCHEMA:
+        // Use a fix schema version for all peers (always in agreement) because stargate waits
```
do you mean that there is happens-before relation between those events?

mpenick replied:
Not sure what you mean here.

tomekl007 replied:
I mean that stargate waits here, but it is related to this https://github.com/stargate/stargate/pull/267#discussion_r510768342 comment, so please disregard this comment

tomekl007 commented on persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java:
```diff
@@ -94,37 +114,179 @@ public void beforeChange(
+
+        // This fix schedules a schema pull for the non-member node and is required because
+        // `StorageService.onChange()` doesn't do this for non-member nodes.
+        MigrationManager.instance.scheduleSchemaPull(endpoint, epState);
```
I see that `updatePeerInfo` returns `Future` and we are not blocking on wait for a previous step. Isn't there a risk for race condition here? Shouldn't we chain the `MigrationManager.instance.scheduleSchemaPull(endpoint, epState);` when the Future returned by the `updatePeerInfo` completes?

mpenick replied:
I thought about this too. I'm replicating what's done in `StorageService`, but maybe that's a long standing bug too? I'll add it, but I'm not sure of the implications of blocking in the gossiper thread...

>

tomekl007 replied:
I don't mean blocking, just chain the subsequent operation.
If we would use `CompletableFuture` we could do smth like
 ```
CompletableFuture.from(() -> StargateSystemKeyspace.updatePeerInfo(
            endpoint, "schema_version", StargateSystemKeyspace.SCHEMA_VERSION, executor);
).thenApply( () -> MigrationManager.instance.scheduleSchemaPull(endpoint, epState));
```

mpenick replied:
Understood. The schema version is constant and not related to pulling the schema from other endpoints. Why do these need to be chained?

tomekl007 commented on persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java:
```diff
@@ -94,37 +114,179 @@ public void beforeChange(
+      return InetAddress.getByName(StorageService.instance.getRpcaddress(endpoint));
+    } catch (UnknownHostException e) {
+      // That should not happen, so log an error, but return the
+      // endpoint address since there's a good change this is right
```
nit `change` -> `chance`

tomekl007 commented on persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java:
```diff
@@ -50,91 +126,152 @@ public void initialize() {}
+  private static ByteBuffer buildColumnValue(String name, InetAddress publicAddress) {
+    switch (name) {
+      case "key":
+        return UTF8Type.instance.decompose("local");
```
nit: `"local"` extract to constant

mpenick replied:
What's the benefit of extracting a constant that's used one time?

tomekl007 commented on stargate-starter/src/main/java/io/stargate/starter/Starter.java:
```diff
@@ -169,13 +169,27 @@
   private boolean useProxyProtocol = false;
 
   @Order(value = 13)
+  @Option(
```
why not add those after the last setting? To keep the ordering of those. Also, `startctl` needs to be modified.

mpenick replied:
The goal was to group the related "proxy" settings together.

mpenick replied:
>

tomekl007 replied:
ok, I see. thx

dougwettlaufer commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 30
- Added 11
           

Complexity increasing per file
==============================
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  11
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  6
- stargate-starter/src/main/java/io/stargate/starter/Starter.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java  21
- persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  9
         

Complexity decreasing per file
==============================
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java  -7
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/StargateSystemKeyspace.java  -13
+ persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java  -9
         

Clones added
============
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  2
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  1
         

Clones removed
==============
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java  -1
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/StargateSystemKeyspace.java  -1
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/EventListenerWrapper.java  -1
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/EventListenerWrapper.java  -1
+ persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6334016&bid=20577739)

dougwettlaufer commented:
![Codacy](https://app.codacy.com/assets/images/favicon.png) Here is an overview of what got changed by this pull request:


```diff

Issues
======
+ Solved 30
- Added 11
           

Complexity increasing per file
==============================
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/ProxyProtocolQueryInterceptor.java  21
- stargate-starter/src/main/java/io/stargate/starter/Starter.java  1
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  11
- persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  9
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  6
         

Complexity decreasing per file
==============================
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/StargateSystemKeyspace.java  -13
+ persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java  -9
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java  -7
         

Clones added
============
- persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/interceptors/DefaultQueryInterceptor.java  2
- persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  1
         

Clones removed
==============
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/StargateSystemKeyspace.java  -1
+ persistence-cassandra-4.0/src/main/java/io/stargate/db/cassandra/impl/interceptors/DefaultQueryInterceptor.java  -1
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/EventListenerWrapper.java  -1
+ persistence-dse-6.8/src/main/java/io/stargate/db/dse/impl/EventListenerWrapper.java  -1
+ persistence-cassandra-3.11/src/main/java/io/stargate/db/cassandra/impl/StargateSystemKeyspace.java  -1
         
```


See the complete overview on [Codacy](https://app.codacy.com/gh/stargate/stargate/pullRequest?prid=6334016&bid=20577739)
